cfg/validate.go

```diff
@@ -28,10 +29,25 @@ func isValidLogRotateConfig(config *LogRotateLoggingConfig) error {
 	return nil
 }
 
+func isValidUrl(u string) (string, error) {
+	decodedUrl, err := url.Parse(u)
+	if err != nil {
+		return "", err
+	}
+	return decodedUrl.String(), nil
+}
+
 // ValidateConfig returns a non-nil error if the config is invalid.
 func ValidateConfig(config *Config) error {
-	if err := isValidLogRotateConfig(&config.Logging.LogRotate); err != nil {
+	var err error
+
+	if err = isValidLogRotateConfig(&config.Logging.LogRotate); err != nil {
 		return fmt.Errorf("error parsing log-rotate config: %w", err)
 	}
+
+	if config.GcsConnection.CustomEndpoint, err = isValidUrl(config.GcsConnection.CustomEndpoint); err != nil {
```

Comment:
Let's not change the parsed config in ValidateConfig. It's mandate is to just return an error if the config is invalid.

Reply:
Valid concern, but I am not sure what will be the best fix in this case. 

We can (1)move it / (2)duplicate the check to rationalize? 
(1) throwing an error during rationalize doesn't look right
(2) I am not in favor of duplicating

Reply:
I am okay with (2) i.e. duplicate the url.Parse call with a TODO to check whether there is any benefit of parsing the rawURL first and then changing the URL back to raw URL.

Reply:
As discussed offline, rawURL parsing is required (added a unit test for it). We will continue with the current approach of doing the check and assignment in the validation method.